server/src/main/java/org/elasticsearch/action/admin/cluster/desiredbalance/TransportGetDesiredBalanceAction.java

```diff
@@ -0,0 +1,111 @@
+import org.elasticsearch.action.support.master.TransportMasterNodeReadAction;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.cluster.block.ClusterBlockLevel;
+import org.elasticsearch.cluster.metadata.DesiredNodes;
+import org.elasticsearch.cluster.metadata.DesiredNodesMetadata;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.allocator.DesiredBalance;
+import org.elasticsearch.cluster.routing.allocation.allocator.DesiredBalanceShardsAllocator;
+import org.elasticsearch.cluster.routing.allocation.allocator.ShardAssignment;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class TransportGetDesiredBalanceAction extends TransportMasterNodeReadAction<DesiredBalanceRequest, DesiredBalanceResponse> {
+
+    private final DesiredBalanceShardsAllocator desiredBalanceShardsAllocator;
+
+    @Inject
+    public TransportGetDesiredBalanceAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        DesiredBalanceShardsAllocator desiredBalanceShardsAllocator
+    ) {
+        super(
+            GetDesiredBalanceAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            DesiredBalanceRequest::new,
+            indexNameExpressionResolver,
+            DesiredBalanceResponse::from,
+            ThreadPool.Names.SAME
+        );
+        this.desiredBalanceShardsAllocator = desiredBalanceShardsAllocator;
+    }
+
+    @Override
+    protected void masterOperation(
+        Task task,
+        DesiredBalanceRequest request,
+        ClusterState state,
+        ActionListener<DesiredBalanceResponse> listener
+    ) throws Exception {
+        DesiredBalance latestDesiredBalance = desiredBalanceShardsAllocator.getDesiredBalance();
+        if (latestDesiredBalance == null) {
+            listener.onFailure(new ResourceNotFoundException("Desired balance not found"));
+            return;
+        }
+        DesiredNodes desiredNodes = DesiredNodesMetadata.fromClusterState(state).getLatestDesiredNodes();
+        Map<String, Map<Integer, DesiredBalanceResponse.DesiredShards>> routingTable = new HashMap<>();
+        Map<ShardId, ShardAssignment> assignments = latestDesiredBalance.assignments();
+        for (Map.Entry<ShardId, ShardAssignment> entry : assignments.entrySet()) {
```

Comment:
I wonder if we should iterate over the routing nodes instead?

We need to keep in mind unassigned shards that might have no balance (in case the index was just created and computation is not yet finished)

Reply:
Yeah, that was one of the earlier ideas. I am going to change it to iterating over the routing nodes.